demo_nodes_cpp/src/services/add_two_ints_client_async.cpp

```diff
@@ -27,51 +28,76 @@ using namespace std::chrono_literals;
 void print_usage()
 {
   printf("Usage for add_two_ints_client app:\n");
-  printf("add_two_ints_client [-t topic_name] [-h]\n");
+  printf("add_two_ints_client [-s service_name] [-h]\n");
   printf("options:\n");
   printf("-h : Print this help function.\n");
   printf("-s service_name : Specify the service name for this client. Defaults to add_two_ints.\n");
 }
 
-int main(int argc, char ** argv)
+class ClientNode : public rclcpp::Node
 {
-  rclcpp::init(argc, argv);
-
-  auto node = rclcpp::Node::make_shared("add_two_ints_client");
-
-  if (rcutils_cli_option_exist(argv, argv + argc, "-h")) {
-    print_usage();
-    return 0;
-  }
+public:
+  explicit ClientNode(const std::string & service_name)
+  : Node("add_two_ints_client")
+  {
+    client_ = create_client<example_interfaces::srv::AddTwoInts>(service_name);
 
-  auto topic = std::string("add_two_ints");
-  if (rcutils_cli_option_exist(argv, argv + argc, "-s")) {
-    topic = std::string(rcutils_cli_get_option(argv, argv + argc, "-s"));
+    // Queue an asynchronous service request that will be sent once `spin` is called on the node.
+    queue_async_request();
   }
-  auto client = node->create_client<example_interfaces::srv::AddTwoInts>(topic);
 
-  while (!client->wait_for_service(1s)) {
-    if (!rclcpp::ok()) {
-      printf("add_two_ints_client was interrupted while waiting for the service. Exiting.\n");
-      return 0;
+  void queue_async_request()
+  {
+    while (!client_->wait_for_service(1s)) {
+      if (!rclcpp::ok()) {
+        printf("add_two_ints_client was interrupted while waiting for the service. Exiting.\n");
+        return;
+      }
+      printf("service not available, waiting again...\n");
     }
-    printf("service not available, waiting again...\n");
+    auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
+    request->a = 2;
+    request->b = 3;
+
+    // We give the async_send_request() method a callback that will get executed once the response
+    // is received.
+    // This way we can return immediately from this method and allow other work to be done by the
+    // executor in `spin` while waiting for the response.
+    using ServiceResponseFuture =
+        rclcpp::client::Client<example_interfaces::srv::AddTwoInts>::SharedFuture;
+    auto response_received_callback = [](ServiceResponseFuture future) {
+        auto result = future.get();
+        printf("Result of add_two_ints: %" PRId64 "\n", result->sum);
+        rclcpp::shutdown();
+      };
+    auto future_result = client_->async_send_request(request, response_received_callback);
   }
 
-  auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
-  request->a = 2;
-  request->b = 3;
+private:
+  rclcpp::client::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client_;
+  rclcpp::timer::TimerBase::SharedPtr timer_;
```

Comment:
we don't need that `timer_` anymore, do we?

Reply:
absolutely right, thanks! b9e57944dee4f88e9030fb1d1cb4233f7e1efd4e